openshift/sandboxed-containers-operator
# KATA-798: openshift controller: use upgradeable condition

**- Description of the problem which is fixed/What is the use case**
There are situations when we don't want OLM to initiate an
update/upgrade of the Operator:
 - during an installation of Kata Containers
 - during an uninstallation of Kata Containers

In these situations the state of Kata on the worker nodes is
in-flight and an update/upgrade could result in inconsistent state

**- What I did**
OLM has built-in support for this and checks if the Operator has a
condition called 'Upgradeable'. OLM will not attempt an update when the
Operator's Upgradeable condition is set to false.
https://olm.operatorframework.io/docs/advanced-tasks/communicating-operator-conditions-to-olm/#upgradeable

We set the condition to *false* when
 - the kataconfig CR was created and we initiated the installation
 - the kataconfig Cr was deleted and we initiated the uninstllation

We set the condition to *true* when
 - an installation has finished
 - an uninstallation has finished

The condition is initialized the first time the reconcile loop is run
and a new kataconfig CR instance was found.

The Upgradeable condition can be overriden by an administrator:
https://olm.operatorframework.io/docs/advanced-tasks/communicating-operator-conditions-to-olm/#overriding-operatorconditions

**- How to verify it**
1. Check that conditions are shown during install/uninstall
2. create a bundle and an index image. Install the operator. add bundle to index image and update operator while installation is ongoing. Update should be denied by OLM.
3. create a bundle and an index image. Install the operator. add bundle to index image and update operator while uninstallation is ongoing. Update should be denied by OLM.

**- Description for the changelog**
Add Upgradeable condition to preven OLM to do updates while install/uninstall is ongoing.

Fixes: https://issues.redhat.com/browse/KATA-798

Signed-off-by: Jens Freimann <jfreimann@redhat.com>
